e2e/internal/e2e/keyexec.go

```diff
@@ -0,0 +1,204 @@
+func RecoverSecretKeyring(t *testing.T) {
+	if err := os.Remove(filepath.Join(HomeDir(), ".singularity/sypgp/pgp-secret")); err != nil {
+		t.Fatalf("Unable to remove secret keyring: %v", err)
+	}
+	if err := os.Rename(backupSypgp, filepath.Join(HomeDir(), ".singularity/sypgp/pgp-secret")); err != nil {
+		t.Fatalf("Unable to rename secret keyring: %v", err)
+	}
+}
+
+// RemoveSecretKeyring will delete your secret keyring :O
+func RemoveSecretKeyring(t *testing.T) {
+	err := os.Remove(filepath.Join(HomeDir(), ".singularity/sypgp/pgp-secret"))
+	if err != nil {
+		t.Fatalf("Unable to remove secret keyring: %v", err)
+	}
+}
+
+// ImportKey will import a key from kpath.
+func ImportKey(t *testing.T, kpath string) ([]byte, error) {
+	LoadEnv(t, &testenv)
+
+	argv := []string{"key", "import", kpath}
+	execKey := exec.Command(testenv.CmdPath, argv...)
+
+	return execKey.CombinedOutput()
+}
+
+// ImportTestKey will take a private key file (kpath) and import it.
+func ImportPrivateKey(t *testing.T, kpath string) ([]byte, error) {
+	s := getImportScript(kpath)
+
+	importScript, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatalf("Unable to create script: %v", err)
+	}
+	defer importScript.Close()
+
+	err = ioutil.WriteFile(importScript.Name(), []byte(s), 0644)
+	if err != nil {
+		t.Fatalf("Unable to write tmp file: %v", err)
+	}
+
+	argv := []string{importScript.Name()}
+	execImport := exec.Command("/usr/bin/expect", argv...)
+
+	return execImport.CombinedOutput()
+}
+
+// ExportPrivateKey will import a private key from kpath.
+func ExportPrivateKey(t *testing.T, kpath string) ([]byte, error) {
+	s := getExportScript(0, kpath, "", "e2etests")
+
+	exportScript, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatalf("Unable to create script: %v", err)
+	}
+	defer exportScript.Close()
+
+	_, err = exportScript.WriteString(s)
+	if err != nil {
+		t.Fatalf("Unable to write to file: %v", err)
+	}
+	err = exportScript.Close()
+	if err != nil {
+		t.Fatalf("Unable to clise file: %v", err)
+	}
+
+	argv := []string{exportScript.Name()}
+	execImport := exec.Command("/usr/bin/expect", argv...)
+
+	return execImport.CombinedOutput()
+}
+
+// RunKeyCmd will run a 'singularty key' command, with any args that are set in commands.
+func RunKeyCmd(t *testing.T, cmdPath string, commands []string, file, stdin string) (string, []byte, error) {
+	argv := []string{"key"}
+	argv = append(argv, commands...)
+
+	if file != "" {
+		argv = append(argv, file)
+	}
+
+	cmd := fmt.Sprintf("%s %s", cmdPath, strings.Join(argv, " "))
+	execKey := exec.Command(cmdPath, argv...)
+
+	stdinRun, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+	defer stdinRun.Close()
+
+	_, err = io.WriteString(stdinRun, stdin)
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	_, err = stdinRun.Seek(0, os.SEEK_SET)
+	if err != nil {
+		t.Fatalf("%v", err)
+	}
+
+	execKey.Stdin = stdinRun
```

Comment:
note that Cmd.Stdin is an io.Reader. A bytes.Buffer is also an io.Reader (also strings.Reader).

This means you can place `stdin` in a bytes.Buffer and you are set. Since `stdin` is a `string` and not a `[]byte`, you can use `strings.NewReader` to obtain an io.Reader.